beartype/typing.py

```diff
@@ -288,3 +280,197 @@
+          >>> my_thing: MyBearProtocol = MyImplementation()
+          >>> isinstance(my_thing, MyBearProtocol)
+          True
+        """
+
+        _abc_inst_check_cache: Dict[Type, bool]
+
+        def __new__(
+            mcls: Type[_TT],
+            name: str,
+            bases: Tuple[Type, ...],
+            namespace: Dict[str, Any],
+            **kw: Any,
+        ) -> _TT:
+            # See <https://github.com/python/mypy/issues/9282>
+            cls = super().__new__(mcls, name, bases, namespace, **kw)  # type: ignore [misc]
+            # Prefixing this class member with "_abc_" is necessary to prevent
+            # it from being considered part of the Protocol. (See
+            # <https://github.com/python/cpython/blob/main/Lib/typing.py>.)
+            cls._abc_inst_check_cache = {}
+
+            return cls
+
+        def __instancecheck__(cls, inst: Any) -> bool:
+            try:
+                # This has to stay *super* tight! Even adding a mere assertion
+                # can add ~50% to the best case runtime!
+                return cls._abc_inst_check_cache[type(inst)]
+            except KeyError:
+                # If you're going to do *anything*, do it here. Try not to
+                # expand the rest of this method if you can avoid it.
+                inst_t = type(inst)
+                bases_pass_muster = True
+
+                for base in cls.__bases__:
+                    if base is cls or base.__name__ in ("Protocol", "Generic", "object"):
+                        continue
+
+                    if not isinstance(inst, base):
+                        bases_pass_muster = False
+                        break
+
+                cls._abc_inst_check_cache[
+                    inst_t
+                ] = bases_pass_muster and cls._check_only_my_attrs(inst)
+
+                return cls._abc_inst_check_cache[inst_t]
+
+    @runtime_checkable
+    class Protocol(_Protocol, Generic, metaclass=_CachingProtocolMeta):
+        """
+        ``@beartype``-compatible (i.e., ``@runtime_checkable``) drop-in
+        replacement for :class:`typing.Protocol` that can lead to significant
+        performance improvements. Uses :class:`_CachingProtocolMeta` to cache
+        :func:`isinstance` check results.
+
+        .. code-block:: python
+          :linenos:
+
+          >>> from abc import abstractmethod
+          >>> from beartype import beartype
+          >>> from beartype.typing import Protocol
+
+          >>> class MyBearProtocol(Protocol):  # <-- runtime-checkable through inheritance
+          ...   @abstractmethod
+          ...   def myfunc(self, arg: int) -> str:
+          ...     pass
+
+          >>> my_thing: MyBearProtocol = MyImplementation()
+          >>> isinstance(my_thing, MyBearProtocol)
+          True
+
+          >>> @beartype
+          ... def do_somthing(thing: MyBearProtocol) -> None:
+          ...   thing.myfunc(0)
+        """
+        __slots__: Union[str, Iterable[str]] = ()
+
```

Comment:
From here on, I create caching versions of the standard library's `Supports` protocols. This might be taking liberties, in which case I can remove them. If they should stay, I'll probably need some guidance on how to overcome the current type name checking test failures.